File: backend/src/main/java/com/shopverse/service/SettingsService.java
package com.shopverse.service;

import com.shopverse.domain.StoreSettings;
import com.shopverse.repository.StoreSettingsRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.LinkedHashMap;
import java.util.Map;

@Service
public class SettingsService {

    private static final long SETTINGS_ID = 1L;

    private final StoreSettingsRepository repository;

    public SettingsService(StoreSettingsRepository repository) {
        this.repository = repository;
    }

    private static final String[] KEYS = {
            "store_name_en", "store_name_ar", "tagline_en", "tagline_ar",
            "announcement_en", "announcement_ar", "announcement_enabled", "shipping_fee",
            "vodafone_number", "instapay_number", "support_phone", "support_email",
            "instagram_url", "facebook_url", "tiktok_url"
    };

    @Transactional
    public Map<String, Object> get() {
        StoreSettings s = ensure();
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("store_name_en", s.getStoreNameEn());
        out.put("store_name_ar", s.getStoreNameAr());
        out.put("tagline_en", s.getTaglineEn());
        out.put("tagline_ar", s.getTaglineAr());
        out.put("announcement_en", s.getAnnouncementEn());
        out.put("announcement_ar", s.getAnnouncementAr());
        out.put("announcement_enabled", s.isAnnouncementEnabled());
        out.put("shipping_fee", s.getShippingFee());
        out.put("vodafone_number", s.getVodafoneNumber());
        out.put("instapay_number", s.getInstapayNumber());
        out.put("support_phone", s.getSupportPhone());
        out.put("support_email", s.getSupportEmail());
        out.put("instagram_url", s.getInstagramUrl());
        out.put("facebook_url", s.getFacebookUrl());
        out.put("tiktok_url", s.getTiktokUrl());
        out.put("updated_at", s.getUpdatedAt());
        return out;
    }

    /** Merge the given patch into the singleton, whitelisted to the known keys (mirrors the mock KEYS list). */
    @Transactional
    public Map<String, Object> update(Map<String, Object> patch) {
        StoreSettings s = ensure();
        for (String key : KEYS) {
            if (patch == null || !patch.containsKey(key)) continue;
            Object value = patch.get(key);
            switch (key) {
                case "announcement_enabled" -> s.setAnnouncementEnabled(Boolean.TRUE.equals(value));
                case "shipping_fee" -> s.setShippingFee(value == null ? BigDecimal.ZERO : new BigDecimal(String.valueOf(value)));
                default -> { if (value != null) setString(s, key, String.valueOf(value)); }
            }
        }
        s.setUpdatedAt(LocalDateTime.now());
        repository.save(s);
        return get();
    }

    @Transactional
    public void reset() {
        repository.deleteAll();
    }

    public StoreSettings ensure() {
        return repository.findById(SETTINGS_ID).orElseGet(() -> {
            StoreSettings s = new StoreSettings();
            s.setId(SETTINGS_ID);
            s.setStoreNameEn("Grounded");
            s.setStoreNameAr("غراوندد");
            s.setTaglineEn("Premium streetwear at honest prices. T-shirts, shirts, pants and more — delivered across Egypt.");
            s.setTaglineAr("ملابس ستريتوير فخمة بأسعار منصفة. تيشيرتات، قمصان وبناطيل — توصيل لجميع مصر.");
            s.setAnnouncementEn("");
            s.setAnnouncementAr("");
            s.setAnnouncementEnabled(false);
            s.setShippingFee(new BigDecimal("80"));
            s.setVodafoneNumber("+20 100 000 0000");
            s.setInstapayNumber("01000000000");
            s.setSupportPhone("+20 100 000 0000");
            s.setSupportEmail("support@grounded.store");
            s.setInstagramUrl("");
            s.setFacebookUrl("");
            s.setTiktokUrl("");
            s.setUpdatedAt(LocalDateTime.now());
            return repository.save(s);
        });
    }

    private void setString(StoreSettings s, String key, String value) {
        switch (key) {
            case "store_name_en" -> s.setStoreNameEn(value);
            case "store_name_ar" -> s.setStoreNameAr(value);
            case "tagline_en" -> s.setTaglineEn(value);
            case "tagline_ar" -> s.setTaglineAr(value);
            case "announcement_en" -> s.setAnnouncementEn(value);
            case "announcement_ar" -> s.setAnnouncementAr(value);
            case "vodafone_number" -> s.setVodafoneNumber(value);
            case "instapay_number" -> s.setInstapayNumber(value);
            case "support_phone" -> s.setSupportPhone(value);
            case "support_email" -> s.setSupportEmail(value);
            case "instagram_url" -> s.setInstagramUrl(value);
            case "facebook_url" -> s.setFacebookUrl(value);
            case "tiktok_url" -> s.setTiktokUrl(value);
            default -> { /* ignored */ }
        }
    }
}

File: backend/src/services/settingsService.js
'use strict'

const db = require('../db')

const SETTINGS_ID = 1

const KEYS = [
  'store_name_en', 'store_name_ar', 'tagline_en', 'tagline_ar',
  'announcement_en', 'announcement_ar', 'announcement_enabled', 'shipping_fee',
  'vodafone_number', 'instapay_number', 'support_phone', 'support_email',
  'instagram_url', 'facebook_url', 'tiktok_url',
]

const UPSERT = `
  INSERT INTO store_settings (id, store_name_en, store_name_ar, tagline_en, tagline_ar,
    announcement_en, announcement_ar, announcement_enabled, shipping_fee, vodafone_number,
    instapay_number, support_phone, support_email, instagram_url, facebook_url, tiktok_url, updated_at)
  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  ON CONFLICT(id) DO UPDATE SET
    store_name_en = excluded.store_name_en,
    store_name_ar = excluded.store_name_ar,
    tagline_en = excluded.tagline_en,
    tagline_ar = excluded.tagline_ar,
    announcement_en = excluded.announcement_en,
    announcement_ar = excluded.announcement_ar,
    announcement_enabled = excluded.announcement_enabled,
    shipping_fee = excluded.shipping_fee,
    vodafone_number = excluded.vodafone_number,
    instapay_number = excluded.instapay_number,
    support_phone = excluded.support_phone,
    support_email = excluded.support_email,
    instagram_url = excluded.instagram_url,
    facebook_url = excluded.facebook_url,
    tiktok_url = excluded.tiktok_url,
    updated_at = excluded.updated_at
`

async function ensure() {
  const row = await db.get('SELECT * FROM store_settings WHERE id = ?', [SETTINGS_ID])
  if (row) return row
  const defaults = {
    store_name_en: 'Grounded',
    store_name_ar: 'غراوندد',
    tagline_en: 'Premium streetwear at honest prices. T-shirts, shirts, pants and more — delivered across Egypt.',
    tagline_ar: 'ملابس ستريتوير فخمة بأسعار منصفة. تيشيرتات، قمصان وبناطيل — توصيل لجميع مصر.',
    announcement_en: '',
    announcement_ar: '',
    announcement_enabled: false,
    shipping_fee: 80,
    vodafone_number: '+20 100 000 0000',
    instapay_number: '01000000000',
    support_phone: '+20 100 000 0000',
    support_email: 'support@grounded.store',
    instagram_url: '',
    facebook_url: '',
    tiktok_url: '',
    updated_at: db.nowIso(),
  }
  await db.run(UPSERT, [
    SETTINGS_ID,
    defaults.store_name_en, defaults.store_name_ar,
    defaults.tagline_en, defaults.tagline_ar,
    defaults.announcement_en, defaults.announcement_ar,
    db.bool(defaults.announcement_enabled), defaults.shipping_fee,
    defaults.vodafone_number, defaults.instapay_number,
    defaults.support_phone, defaults.support_email,
    defaults.instagram_url, defaults.facebook_url, defaults.tiktok_url,
    defaults.updated_at,
  ])
  return db.get('SELECT * FROM store_settings WHERE id = ?', [SETTINGS_ID])
}

function toMap(row) {
  return {
    store_name_en: row.store_name_en,
    store_name_ar: row.store_name_ar,
    tagline_en: row.tagline_en,
    tagline_ar: row.tagline_ar,
    announcement_en: row.announcement_en,
    announcement_ar: row.announcement_ar,
    announcement_enabled: row.announcement_enabled,
    shipping_fee: row.shipping_fee,
    vodafone_number: row.vodafone_number,
    instapay_number: row.instapay_number,
    support_phone: row.support_phone,
    support_email: row.support_email,
    instagram_url: row.instagram_url,
    facebook_url: row.facebook_url,
    tiktok_url: row.tiktok_url,
    updated_at: row.updated_at,
  }
}

async function get() {
  return toMap(await ensure())
}

async function update(patch) {
  const row = await ensure()
  const next = { ...row }
  for (const key of KEYS) {
    if (patch === null || patch === undefined || !(key in patch)) continue
    const value = patch[key]
    if (key === 'announcement_enabled') {
      next.announcement_enabled = value === true || value === 'true' || value === 1
    } else if (key === 'shipping_fee') {
      const n = Number(value)
      next.shipping_fee = Number.isNaN(n) ? 0 : n
    } else if (value !== null && value !== undefined) {
      next[key] = String(value)
    }
  }
  next.updated_at = db.nowIso()
  await db.run(UPSERT, [
    SETTINGS_ID,
    next.store_name_en, next.store_name_ar,
    next.tagline_en, next.tagline_ar,
    next.announcement_en, next.announcement_ar,
    db.bool(next.announcement_enabled), next.shipping_fee,
    next.vodafone_number, next.instapay_number,
    next.support_phone, next.support_email,
    next.instagram_url, next.facebook_url, next.tiktok_url,
    next.updated_at,
  ])
  return get()
}

async function reset() {
  await db.run('DELETE FROM store_settings')
}

module.exports = { ensure, get, update, reset, KEYS }
